cluster/connector.go

```diff
@@ -0,0 +1,187 @@
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type Connector struct {
+	node  *core.IpfsNode
+	api   iface.CoreAPI
+	peers func(ctx context.Context) []*api.ID
+}
+
+func NewConnector(node *core.IpfsNode, peers func(ctx context.Context) []*api.ID) (*Connector, error) {
+	capi, err := coreapi.NewCoreAPI(node)
+	if err != nil {
+		return nil, err
+	}
+	return &Connector{
+		node:  node,
+		api:   capi,
+		peers: peers,
+	}, nil
+}
+
+func (c *Connector) ID(context.Context) (*api.IPFSID, error) {
+	var addrs []api.Multiaddr
+	for _, addr := range c.node.PeerHost.Addrs() {
+		addrs = append(addrs, api.Multiaddr{Multiaddr: addr})
+	}
+	return &api.IPFSID{
+		ID:        c.node.Identity,
+		Addresses: addrs,
+	}, nil
+}
+
+func (c *Connector) SetClient(client *rpc.Client) {
+	// noop
+}
+
+func (c *Connector) Shutdown(ctx context.Context) error {
+	// noop
+	return nil
+}
+
+// @todo handle maxDepth
+func (c *Connector) Pin(ctx context.Context, cid icid.Cid, maxDepth int) error {
+	return c.api.Pin().Add(ctx, path.New(cid.String()))
+}
+
+func (c *Connector) Unpin(ctx context.Context, cid icid.Cid) error {
+	return c.api.Pin().Rm(ctx, path.New(cid.String()))
+}
+
+func (c *Connector) PinLsCid(ctx context.Context, cid icid.Cid) (api.IPFSPinStatus, error) {
+	pins, err := c.node.Pinning.CheckIfPinned(cid)
+	if err != nil {
+		return api.IPFSPinStatusError, err
+	}
+	if len(pins) == 0 {
+		return api.IPFSPinStatusError, fmt.Errorf("invalid pin check result")
+	}
+	return c.pinModeToStatus(pins[0].Mode), nil
+}
+
+func (c *Connector) PinLs(ctx context.Context, typeFilter string) (map[string]api.IPFSPinStatus, error) {
+	pins, err := c.api.Pin().Ls(ctx, c.pinFilterToOption(typeFilter))
+	if err != nil {
+		return nil, err
+	}
+	statusMap := make(map[string]api.IPFSPinStatus)
+	for _, p := range pins {
+		mode, ok := pin.StringToMode(p.Type())
+		if !ok {
+			continue
+		}
+		statusMap[p.Path().String()] = c.pinModeToStatus(mode)
+	}
+	return statusMap, nil
+}
+
+func (c *Connector) ConnectSwarms(ctx context.Context) error {
+	for _, p := range c.peers(ctx) {
+		log.Debugf("cluster dialing %s", p.ID.Pretty())
+		var addrs []ma.Multiaddr
+		for _, addr := range p.Addresses {
+			addrs = append(addrs, addr.Multiaddr)
+		}
+		err := c.api.Swarm().Connect(ctx, corepeer.AddrInfo{
+			ID:    p.ID,
+			Addrs: addrs,
+		})
+		if err != nil {
+			return err
+		}
+		log.Debugf("cluster connected to %s")
+	}
+	return nil
+}
+
+func (c *Connector) SwarmPeers(ctx context.Context) ([]peer.ID, error) {
+	conns, err := c.api.Swarm().Peers(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var peers []peer.ID
+	for _, c := range conns {
+		peers = append(peers, c.ID())
+	}
+	return peers, nil
+}
+
+func (c *Connector) ConfigKey(keypath string) (interface{}, error) {
+	return c.node.Repo.GetConfigKey(keypath)
+}
+
+func (c *Connector) RepoStat(ctx context.Context) (*api.IPFSRepoStat, error) {
+	stat, err := corerepo.RepoStat(ctx, c.node)
```

Comment:
To check: we call ipfs with `size-only=true`. https://github.com/ipfs/ipfs-cluster/blob/master/ipfsconn/ipfshttp/ipfshttp.go#L600

This makes the RepoStat call WAY faster (no full node count).

Reply:
👌